### PR TITLE
Content changes for automatically declined rejected applications

### PR DESCRIPTION
**Pull request by SamJamCul**
## Context

Changes that implement the design requirements set out here - https://bat-design-history.netlify.app/manage-teacher-training-applications/showing-when-an-offer-will-be-automatically-declined/

Some of the changes requested have been completed in another [PR](https://github.com/DFE-Digital/apply-for-teacher-training/pull/4394)

## Changes proposed in this pull request

- Added declined by default to application card for offered state
- Update rejected by default display for application list
- Change rejected by default text to application choice header

## Guidance to review

For the application list, look at the text next to applications in the offered and received/interviewing states
image

image

For the application choice header, select a candidate in the offered state

image

## Link to Trello card

https://trello.com/c/gPo8FHgx/3580-content-changes-for-automatically-declined-rejected-applications

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/components/provider_interface/application_card_component.html.erb [hunk 1]
```diff
@@ -21,10 +21,13 @@
     </div>
 
     <div class="govuk-grid-column-one-third">
-      <dl class="app-application-card__list app-application-card__list--secondary">
+      <dl class="app-application-card__list app-applicationq-card__list--secondary">
```
Comment (despo) [inline app/components/provider_interface/application_card_component.html.erb:24]:
```suggestion
      <dl class="app-application-card__list app-application-card__list--secondary">
```

Comment (SamJamCul) [reply app/components/provider_interface/application_card_component.html.erb:24]: Good spot!

File: app/components/provider_interface/application_card_component.rb [hunk 2]
```diff
@@ -18,7 +18,21 @@ def initialize(application_choice:)
...
-        "#{days_until(Date.current + days_left_to_respond).capitalize} to respond"
+        if days_left_to_respond.zero?
+          'Last day to make decision'
+        else
+          "#{days_until(Date.current + days_left_to_respond).capitalize} to make decision"
+        end
+      end
+    end
+
+    def decision_by_date_text
```
Comment (despo) [inline app/components/provider_interface/application_card_component.rb:29]: Maybe provide a bit more context with the method name, e.g. `candidate_days_to_respond_text`?

File: app/components/provider_interface/application_card_component.rb [hunk 3]
```diff
@@ -18,7 +18,21 @@ def initialize(application_choice:)
...
+      if (days_left_to_respond = application_choice.days_until_decline_by_default)
+        if days_left_to_respond.zero?
```
Comment (despo) [inline app/components/provider_interface/application_card_component.rb:31]: Reversing this makes it easier to read

e.g. 
```ruby
if days_left_to_respond.any?
  # do something
else
  # do something else
end

File: app/components/provider_interface/application_choice_header_component.html.erb [hunk 4]
```diff
@@ -16,10 +16,12 @@
             <% end %>
           </h2>
           <p class="govuk-body">
-            <% if time_is_today_or_tomorrow?(application_choice.reject_by_default_at) -%>
-              You have until <%= time_today_or_tomorrow(application_choice.reject_by_default_at) %> to respond to this application. Otherwise it will be automatically rejected.
+            <% if application_choice.reject_by_default_at.today? -%>
+              This application will be automatically rejected at the end of today (<%= time_today_or_tomorrow(application_choice.reject_by_default_at) %>) if you do not make a decision.
+            <% elsif time_is_tomorrow?(application_choice.reject_by_default_at) -%>
+              This application will be automatically rejected at the end of tomorrow (<%= time_today_or_tomorrow(application_choice.reject_by_default_at) %>) if you do not make a decision.
```
Comment (despo) [inline app/components/provider_interface/application_choice_header_component.html.erb:22]: Can we extract the conditions in a method which decides if we should print `today` or `tomorrow`, as the rest of the text is the same?

```suggestion
            <% if application_choice.reject_by_default_at.today? ||  time_is_tomorrow?(application_choice.reject_by_default_at) -%>
              This application will be automatically rejected at the end of <%= some_method_that_returns_today_or_tomorrow %> (<%= time_today_or_tomorrow(application_choice.reject_by_default_at) %>) if you do not make a decision.
```

File: app/components/provider_interface/application_choice_header_component.html.erb [hunk 5]
```diff
@@ -36,10 +38,12 @@
             Make a decision
           </h2>
           <p class="govuk-body">
-            <% if time_is_today_or_tomorrow?(application_choice.reject_by_default_at) -%>
-              You have until <%= time_today_or_tomorrow(application_choice.reject_by_default_at) %> to respond to this application. Otherwise it will be automatically rejected.
+            <% if application_choice.reject_by_default_at.today? -%>
+              This application will be automatically rejected at the end of today (<%= time_today_or_tomorrow(application_choice.reject_by_default_at) %>) if you do not make a decision.
+            <% elsif time_is_tomorrow?(application_choice.reject_by_default_at) -%>
+              This application will be automatically rejected at the end of tomorrow (<%= time_today_or_tomorrow(application_choice.reject_by_default_at) %>) if you do not make a decision.
```
Comment (despo) [inline app/components/provider_interface/application_choice_header_component.html.erb:44]: Same as above
https://github.com/DFE-Digital/apply-for-teacher-training/pull/4484/files#r612361779

File: app/models/application_choice.rb [hunk 6]
```diff
@@ -64,17 +64,23 @@ def recruitment_cycle
...
-    if respond_to?(:pg_days_left_to_respond)
-      # pre-computed by sorting query
-      return pg_days_left_to_respond
-    end
-
-    if status == 'awaiting_provider_decision'
+    # if respond_to?(:pg_days_left_to_respond)
+    #   # pre-computed by sorting query
+    #   return pg_days_left_to_respond
+    # end
```
Comment (despo) [inline app/models/application_choice.rb:70]: can we remove the commented out code please.

File: app/models/application_choice.rb [hunk 7]
```diff
@@ -64,17 +64,23 @@ def recruitment_cycle
...
+    if ApplicationStateChange::DECISION_PENDING_STATUSES.include?(status.to_sym)
       rbd = reject_by_default_at
       ((rbd - Time.zone.now) / 1.day).floor if rbd && rbd > Time.zone.now
     end
   end
 
+  def days_until_decline_by_default
+    if status == 'offer'
```
Comment (despo) [inline app/models/application_choice.rb:78]:
```suggestion
    if offer?
```

File: app/models/application_choice.rb [hunk 8]
```diff
@@ -64,17 +64,23 @@ def recruitment_cycle
...
+      dbd = decline_by_default_at
+      ((dbd - Time.zone.now) / 1.day).floor if dbd && dbd > Time.zone.now
```
Comment (despo) [inline app/models/application_choice.rb:80]: Can we merge the conditions, as it adds a bit of confusion getting in an if statement that will only return something if another condition is satisfied.

e.g.
```ruby
dbd = decline_by_default_at
if offer? && dbd && dbd > Time.zone.now
    ((dbd - Time.zone.now) / 1.day).floor
end
```

File: spec/components/provider_interface/application_card_component_spec.rb [hunk 9]
```diff
@@ -156,19 +156,64 @@
...
+    around do |example|
+      Timecop.freeze(Time.zone.local(2020, 6, 1, 12, 30, 0)) { example.run }
+    end
+
+    let(:application_choice) do
+      create(
+        :application_choice,
+        :with_offer,
+        decline_by_default_at: dbd,
+        updated_at: Time.zone.parse('2020-06-01T09:05:00+01:00'),
```
Comment (despo) [inline spec/components/provider_interface/application_card_component_spec.rb:185]: why is this needed?

File: spec/helpers/view_helper_spec.rb [hunk 10]
```diff
@@ -160,6 +160,20 @@
     end
   end
 
+  describe '#time_is_tomorrow' do
+    it 'is true for tomorrow' do
+      expect(helper.time_is_tomorrow?(Time.zone.now + 1.day)).to be true
+    end
+
+    it 'is not true for a time in 49 hours' do
```
Comment (despo) [inline spec/helpers/view_helper_spec.rb:168]:
```suggestion
    it 'is not true for a time after tomorrow' do
```

File: app/components/provider_interface/application_card_component.rb (hunk 2 above)
Comment (SamJamCul) [reply app/components/provider_interface/application_card_component.rb:29]: 👍

File: app/components/provider_interface/application_card_component.rb (hunk 3 above)
Comment (SamJamCul) [reply app/components/provider_interface/application_card_component.rb:31]: 👍

Comment (SamJamCul) [reply app/components/provider_interface/application_card_component.rb:31]: Changed it to `.positive?` as it's an integer

File: app/models/application_choice.rb (hunk 6 above)
Comment (SamJamCul) [reply app/models/application_choice.rb:70]: 👍

File: app/models/application_choice.rb (hunk 7 above)
Comment (SamJamCul) [reply app/models/application_choice.rb:78]: 👍

File: app/models/application_choice.rb (hunk 8 above)
Comment (SamJamCul) [reply app/models/application_choice.rb:80]: 👍

File: spec/helpers/view_helper_spec.rb (hunk 10 above)
Comment (SamJamCul) [reply spec/helpers/view_helper_spec.rb:168]: 👍

File: spec/components/provider_interface/application_card_component_spec.rb (hunk 9 above)
Comment (SamJamCul) [reply spec/components/provider_interface/application_card_component_spec.rb]: Removed!

File: app/components/provider_interface/application_choice_header_component.html.erb (hunk 4 above)
Comment (SamJamCul) [reply app/components/provider_interface/application_choice_header_component.html.erb]: 👍

File: app/frontend/styles/_application-card.scss [hunk 11]
```diff
@@ -21,13 +21,13 @@
 
   .govuk-grid-column-one-third {
     @include govuk-media-query($from: tablet) {
-      width: 37%;
+      width: 40%;
```
Comment (despo) [inline app/frontend/styles/_application-card.scss:24]: As this is used across the app, is there any reason why it is changed as part of this PR?

Comment (cwrw) [reply app/frontend/styles/_application-card.scss:24]: I agree, maybe we can add an override to this under a new class name, similar to this, and append that to the application page specifically?

Comment (SamJamCul) [reply app/frontend/styles/_application-card.scss:24]: This change is in `_application-card.scss`, and under the `.app-application-card` class. It will only affect elements that are already classified as application cards. It's going to be a bit tougher to try and scope this only for specific instances of the application card component.

Comment (despo) [reply app/frontend/styles/_application-card.scss:24]: Sorry about the mirsunderstanding. The generic component name made me think this was reusable across a number of forms!

Comment (cwrw) [reply app/frontend/styles/_application-card.scss:24]: ah oops awesome, this is exactly what we need 👍

**Review by cwrw — COMMENTED** [review]
looking good 👍 , could you also address the following point from the [design history](https://bat-design-history.netlify.app/manage-teacher-training-applications/showing-when-an-offer-will-be-automatically-declined/#showing-on-the-application-list-when-an-offer-will-be-automatically-declined):
* The group heading will be changed from “Waiting for candidate action“ to “Waiting for candidate to respond to offer”. 
I think this was missed in the trello card

**Comment by SamJamCul** [comment]
> 
> 
> looking good 👍 , could you also address the following point from the [design history](https://bat-design-history.netlify.app/manage-teacher-training-applications/showing-when-an-offer-will-be-automatically-declined/#showing-on-the-application-list-when-an-offer-will-be-automatically-declined):
> 
>     * The group heading will be changed from “Waiting for candidate action“ to “Waiting for candidate to respond to offer”.
>       I think this was missed in the trello card

It turned out this had been changed in a previous commit, so didn't need to be actioned - we did spend some time looking for it though, as it was nowhere to be found 🤦‍♂️

**Comment by SamJamCul** [comment]
> looking good 👍 , could you also address the following point from the [design history](https://bat-design-history.netlify.app/manage-teacher-training-applications/showing-when-an-offer-will-be-automatically-declined/#showing-on-the-application-list-when-an-offer-will-be-automatically-declined):
> 
> * The group heading will be changed from “Waiting for candidate action“ to “Waiting for candidate to respond to offer”.
>   I think this was missed in the trello card


@cwrw Done, good find

**Review by despo — APPROVED** [review]
👍🏼

**Review by cwrw — APPROVED** [review]
Looks good, thanks for addressing comments. Happy to :shipit: after the last tests are fixed

File: app/components/provider_interface/application_choice_header_component.html.erb [hunk 12]
```diff
@@ -17,9 +17,9 @@
           </h2>
           <p class="govuk-body">
             <% if time_is_today_or_tomorrow?(application_choice.reject_by_default_at) -%>
-              You have until <%= time_today_or_tomorrow(application_choice.reject_by_default_at) %> to respond to this application. Otherwise it will be automatically rejected.
```
Comment (benswannack) [inline app/components/provider_interface/application_choice_header_component.html.erb:20]: Is `time_today_or _tomorrow` still used anywhere? It can potentially be removed from view helper now we've moved towards showing the date and time.

Comment (SamJamCul) [reply app/components/provider_interface/application_choice_header_component.html.erb:20]: I think this may have been addressed in the latest changes pushed

File: spec/helpers/view_helper_spec.rb [hunk 13]
```diff
@@ -99,15 +99,15 @@
       expect(helper.time_is_today_or_tomorrow?(Time.zone.now - 24.hours)).to be false
     end
 
-    it 'is not true for a time in 49 hours' do
-      expect(helper.time_is_today_or_tomorrow?(Time.zone.now - 49.hours)).to be false
+    it 'is not true for a time after tomorrow' do
```
[the remark this answers is not in the document]
Comment (SamJamCul) [reply spec/helpers/view_helper_spec.rb:102]: 👍

File: spec/components/provider_interface/application_choice_header_component_spec.rb [hunk 14]
```diff
@@ -21,7 +21,7 @@
           expect(result.css('.govuk-button').first.text).to eq('Set up interview')
           expect(result.css('.govuk-button').last.text).to eq('Make decision')
           expect(result.css('.govuk-inset-text').text).to include(
-            'You have until 12pm (midday) tomorrow to respond to this application. Otherwise it will be automatically rejected.',
+            'This application will be automatically rejected at the end of tomorrow (12pm (midday) tomorrow) if you do not make a decision.',
```
[the remark this answers is not in the document]
Comment (SamJamCul) [reply spec/components/provider_interface/application_choice_header_component_spec.rb]: 👍

File: app/components/provider_interface/application_choice_header_component.html.erb (hunk 12 above)
Comment (benswannack) [reply app/components/provider_interface/application_choice_header_component.html.erb:20]: Ah, I expected for us to be able to delete the method after this change.

Have we missed a matching change here? https://github.com/DFE-Digital/apply-for-teacher-training/pull/4484/files#diff-84a45f10edd6890b4815db97d0541ba6d809a6009079d3fcbaef1e3cbeb281e8R51
